### PR TITLE
fix: fix ack trackers data race

### PR DESCRIPTION
**Pull request by KennyChenFight**
We need to add mutex lock to prevent data race

### CODE REVIEW

File: pkg/source/pulsar.go [hunk 1]
```diff
@@ -286,9 +286,56 @@ func newAckTracker(size uint) *ackTracker {
...
+}
+
+func (a *ackTrackers) tryAdd(msg pulsar.MessageID) (ok bool) {
+	key := a.key(msg)
+	a.mu.RLock()
+	_, ok = a.trackers[key]
+	a.mu.RUnlock()
+	if !ok {
+		a.mu.Lock()
+		a.trackers[key] = newAckTracker(uint(msg.BatchSize()))
```
Comment (rueian) [inline pkg/source/pulsar.go:320]: `_, ok = a.trackers[key]` condition should be rechecked after acquiring the lock.

Comment (KennyChenFight) [reply pkg/source/pulsar.go:320]: indeed

Comment (KennyChenFight) [reply pkg/source/pulsar.go:320]: I think maybe we could use sync.Map to handle this situation.
In our case, we have a read-heavy and write-light scenario. Each batch message involves only one write operation and a final delete key action. There is no need to modify the value.

File: pkg/source/pulsar.go [hunk 2]
```diff
@@ -286,9 +286,41 @@ func newAckTracker(size uint) *ackTracker {
...
+func newAckTrackers() *ackTrackers {
+	return &ackTrackers{}
+}
+
+func (a *ackTrackers) key(msg pulsar.MessageID) string {
+	return fmt.Sprintf("%d:%d", msg.LedgerID(), msg.EntryID())
+}
+
+func (a *ackTrackers) tryAdd(msg pulsar.MessageID) bool {
+	_, loaded := a.trackers.LoadOrStore(a.key(msg), newAckTracker(uint(msg.BatchSize())))
```
Comment (rueian) [inline pkg/source/pulsar.go:311]: But doing newAckTracker(uint(msg.BatchSize()) every time won’t be too costly?

Comment (KennyChenFight) [reply pkg/source/pulsar.go]: good point

**Review by rueian — APPROVED** [review]

**Review by benjamin99 — APPROVED** [review]

**Pull request merged**